gnc/sub8_controller/nodes/movement_validity.cpp

```diff
@@ -0,0 +1,81 @@
+#include <ros/ros.h>
+#include <geometry_msgs/PoseStamped.h>
+#include "sub8_msgs/WaypointValidity.h"
+#include <nav_msgs/OccupancyGrid.h>
+
+#include <opencv2/core/core.hpp>
+#include "opencv2/opencv.hpp"
+
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <functional>
+/*
+* Goal: This is going to need to check whether the current waypoint is valid.
+* Check Z coordinate of the waypoint against the current depth of the sub to determine-
+* -whether this is attempting to move outside the water or not.
+* We are also going to eventually want to check whether this is going to be moving-
+* -into a wall or not.
+*/
+
+using namespace std;
+using namespace geometry_msgs;
+
+// Todo: Implement check for ogrid
+
+class WaypointChecker
+{
+private:
+  ros::NodeHandle nh_;
+  ros::ServiceServer validity_service_;
+  nav_msgs::OccupancyGridConstPtr ogrid_map_;
+  ros::Subscriber sub_;
+public:
+
+  WaypointChecker() : nh_(ros::this_node::getName())
+  {
+    sub_ = nh_.subscribe<nav_msgs::OccupancyGrid>("/ogridgen/ogrid", 1,
+                                                  boost::bind(&WaypointChecker::ogrid_callback, this, _1));
+    validity_service_ = nh_.advertiseService<sub8_msgs::WaypointValidity::Request, sub8_msgs::WaypointValidity::Response>(
+        "is_waypoint_valid", boost::bind(&WaypointChecker::is_waypoint_valid, this, _1, _2));
+  }
+
+  // TODO: What the error was
+  bool is_waypoint_valid(sub8_msgs::WaypointValidity::Request &req, sub8_msgs::WaypointValidity::Response &resp)
```

Comment:
I think the approach for this should be different. What if this was a library imported into C3 to reject all bad waypoints instead of being required to have a server running